sling/nlp/embedding/embedding-model.cc

```diff
@@ -0,0 +1,250 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "sling/nlp/embedding/embedding-model.h"
+
+#include "sling/myelin/builder.h"
+#include "sling/myelin/gradient.h"
+#include "sling/util/random.h"
+
+namespace sling {
+namespace nlp {
+
+using namespace myelin;
+
+void MikolovFlow::Build() {
+  BuildModel();
+  BuildLayer0();
+  BuildLayer1();
+  BuildLayer0Back();
+}
+
+void MikolovFlow::BuildModel() {
+  W0 = AddWeights("W0", DT_FLOAT, {inputs, dims});
+  W1 = AddWeights("W1", DT_FLOAT, {outputs, dims});
+  W0->set_random();
+}
+
+void MikolovFlow::BuildLayer0() {
+  layer0 = AddFunction("layer0");
+  FlowBuilder tf(this, layer0);
+
+  fv = tf.Var("features", DT_INT32, {1, in_features});
+  hidden = tf.Name(tf.GatherAvg(W0, fv), "hidden");
+}
+
+void MikolovFlow::BuildLayer1() {
+  layer1 = AddFunction("layer1");
+  FlowBuilder tf(this, layer1);
+
+  // Inputs.
+  alpha = tf.Var("alpha", DT_FLOAT, {});
+  label = tf.Var("label", DT_FLOAT, {1, 1});
+  target = tf.Var("target", DT_INT32, {1, out_features});
+  error = tf.Var("error", DT_FLOAT, {dims});
+  l1_l0 = tf.Instance(layer0);
+  auto *h = tf.Ref(l1_l0, hidden);
+
+  // Output.
+  bool single = out_features == 1;
+  auto *embed = single ? tf.Gather(W1, target) : tf.GatherAvg(W1, target);
+  auto *output = tf.Dot(embed, h, dims);
+
+  // Loss.
+  loss = tf.Name(tf.Sub(label, tf.Sigmoid(output)), "loss");
```

Comment:
Actually, this is not the loss (or loss_grad) but the likelihood, since you are doing gradient ascent later on.